### PR TITLE
feat: implement package

### PR DESCRIPTION
**Pull request by vitaligi**
some more work is needed on the documentation of the functions, it will be added later

| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✔                                                                        |
| Chore            | ✖                                                                       |

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -1,3 +1,149 @@
...
+function tileWidth(zoom: Zoom, referenceTileGrid: TileGrid): number {
+  return (
+    (referenceTileGrid.boundingBox.east - referenceTileGrid.boundingBox.west) / (referenceTileGrid.numberOfMinLevelTilesX * SCALE_FACTOR ** zoom)
+  );
+}
+
+function* tilesGenerator(limits: [number, number, number, number], zoom: Zoom, metatile: number): Generator<Tile, undefined, undefined> {
+  for (let y = limits[1]; y <= limits[3]; y++) {
+    for (let x = limits[0]; x <= limits[2]; x++) {
+      yield { x: x, y: y, z: zoom, metatile: metatile };
```
Comment (CptSchnitz) [inline src/index.ts]: 1. what is the reasoning behind returning the metatile if its unchanged and unused?
2. using the shortened method for property value will make it more readable https://attacomsian.com/blog/javascript-object-property-shorthand

File: src/index.ts [hunk 2]
```diff
@@ -1,3 +1,149 @@
...
+    }
+  }
+
+  return;
+}
+
+export function boundingBoxToTiles(
+  bbox: BoundingBox,
+  zoom: Zoom,
+  metatile = 1,
```
Comment (CptSchnitz) [inline src/index.ts]: what does the number of the metatile represents?

File: src/index.ts [hunk 3]
```diff
@@ -1,3 +1,149 @@
...
+  );
+}
+
+function tileWidth(zoom: Zoom, referenceTileGrid: TileGrid): number {
+  return (
+    (referenceTileGrid.boundingBox.east - referenceTileGrid.boundingBox.west) / (referenceTileGrid.numberOfMinLevelTilesX * SCALE_FACTOR ** zoom)
+  );
+}
+
+function* tilesGenerator(limits: [number, number, number, number], zoom: Zoom, metatile: number): Generator<Tile, undefined, undefined> {
```
Comment (CptSchnitz) [inline src/index.ts]: create a type for [number, number, number, number]. This structure can be confused with Bbox easily

File: src/index.ts [hunk 4]
```diff
@@ -1,3 +1,149 @@
...
+    // the value is validated before so this should be unreachable
+    throw new Error('zoom level is not part of the given well known scale set');
+  }
+
+  let matchingZoom: Zoom | undefined;
+  targetTileGrid.wellKnownScaleSet.scaleDenominators.forEach((targetScaleDenominator, targetZoom) => {
+    if (targetScaleDenominator === scale) {
+      matchingZoom = targetZoom;
+    }
+  });
```
Comment (CptSchnitz) [inline src/index.ts]: consider using for..of loop so you can return a result once it was found instead of looping all over.

**Review by CptSchnitz — CHANGES_REQUESTED** [review]
1. When adding docs to the functions, add it using typedoc so we can generate a page for it https://typedoc.org/example/
2. consider adding tests that check negative values for latitude and longtitude.

File: src/index.ts (hunk 2 above)
Comment (vitaligi) [reply src/index.ts]: it is the number of tiles in both x and y axis the metatile holds

File: src/index.ts (hunk 1 above)
Comment (vitaligi) [reply src/index.ts]: 1. it should be part of the tile because it affects the x and y numbering. for example in the case you received a tile you need to know which metatile number was involved in it's creation
2 . ok

File: src/index.ts (hunk 4 above)
Comment (vitaligi) [reply src/index.ts]: done

File: src/index.ts (hunk 3 above)
Comment (vitaligi) [reply src/index.ts]: done

**Review by CptSchnitz — APPROVED** [review]
